cachi2/core/package_managers/rubygems/main.py

```diff
@@ -0,0 +1,95 @@
+import logging
+from typing import Optional
+
+from cachi2.core.models.input import Request
+from cachi2.core.models.output import EnvironmentVariable, ProjectFile, RequestOutput
+from cachi2.core.models.sbom import Component
+from cachi2.core.package_managers.rubygems.parser import (
+    BundlerDependency,
+    GemDependency,
+    GitDependency,
+    PathDependency,
+)
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+def fetch_rubygems_source(request: Request) -> RequestOutput:
+    """Resolve and process all rubygems packages."""
+    components: list[Component] = []
+    environment_variables: list[EnvironmentVariable] = []
+    project_files: list[ProjectFile] = []
+
+    for package in request.packages:
+        package_dir = request.source_dir.join_within_root(package.path)
+        _resolve_rubygems(package_dir=package_dir, output_dir=request.output_dir)
+
+    return RequestOutput.from_obj_list(
+        components=components,
+        environment_variables=environment_variables,
+        project_files=project_files,
+    )
+
+
+def _resolve_rubygems(package_dir: RootedPath, output_dir: RootedPath) -> list[Component]:
+    """
+    Process a request for a single rubygems package.
+
+    TODO:
+    1. Parse Gemfile.lock
+    2. Get project metadata
+    3. Download dependencies
+    4. Generate SBOM components
+    """
+    return NotImplemented
+
+
+def _get_package_metadata(
+    package_dir: RootedPath,
+    path_dependencies: list[PathDependency],
+) -> tuple[str, Optional[str]]:
+    """TODO."""
+    return NotImplemented
+
+
+def _get_repository_name(package_dir: RootedPath) -> str:
+    """TODO."""
+    return NotImplemented
+
+
+def _download_dependencies(output_dir: RootedPath, dependencies: list[BundlerDependency]) -> None:
```

Comment:
Are you by any chance a fan of Haskell? :grin: 

Might as well take it a step further:

```python
type IO[T] = T

def _download_dependencies(...) -> IO[None]:
    ...
```

In all seriousness, I think introducing effect types into Python is a step too far. It would make this codebase feel alien to Python devs.

Reply:
That's a good one! Given that Python drifts towards becoming a poor man's ML dialect among other things (I am looking at you `match`) this might be the right decision!

TBH all the fuss with type annotations does feel a little alien to me as a Python dev: the same result could be achieved with other means without introducing that much visual noise. However annotations do give some helpful hints sometimes, and I think they give higher-quality hints when types are well-refined. That is why you'll see me commenting on types and names quite often. I would rarely if ever block on such comments though.

The main benefit here would be to immediately tell the reader what to expect. This is definitely a slippery slope which may end up with checked exceptions, so please consider this an out-loud thought experiment.

Reply:
> the same result could be achieved with other means without introducing that much visual noise.

Like function and parameter naming?

In addition to being hints for the reader, the value of type hints is static type-checking (which I find immensely helpful).

A type hint like `SideEffect` would only serve the former, it has no benefits for type-checking. IMO the names `_download_dependencies, output_dir` make it clear enough what to expect from this function

Reply:
Fair enough, my original concern was with effective overloading of `None` symbol to also hint on side-effects. The combination of names indeed implies that dependencies will be not only downloaded, but also probably saved to some directory.